### ISSUE TITLE
Lightroom Integration

### ISSUE THREAD

**Issue by ajamess** [SanjaySoniLV/ProjectKestrel#1]
This is an extremely valuable project, but I think it needs some tooling integration to be of use for most photographers. Have you considered a lightroom metadata plugin? You could pretty easily write an integration to take the raw data you output in terms of species, confidence, quality, similarity, etc, and write it to the lightroom database alongside the file location.

This way a photographer could:

1) Import photos into lightroom
2) Launch your tool to run ML detection on them
3) Write output of the run to the lightroom database entry associated which each of those images (or as an image keyword)
4) Allow the photographer to sort and cull in lightroom based on that metadata 

Another way to do this would be to write that metadata into XMP files associated with each raw file, if you didn't want to futz around with the Lightroom SDK.

**Comment by ajamess** [comment]
I did a test run over 8500 images. Using [LR/Transporter](https://www.photographers-toolbox.com/products/lrtransporter.php) I was able to quickly rate each image based on the quality detected by your tool, and then extend the keywords to include the detected species. In general, the scene detection is also really good, but is harder to import in this manual way.

<img width="1147" height="955" alt="Image" src="https://github.com/user-attachments/assets/96b841c7-9d62-4034-8c09-670fa1510235" />

I think with a bit of end to end friction reduction in the process, this could be an extremely valuable tool for bird photographers. I think you should consider packaging this up into a lightroom plugin, and then have a way to extend the image database to include the metadata here.  For example, the scene is very valuable information to quickly "bucket" shots into groups, but it is not possible to meaningfully import in lightroom without extending the database.

This way, you could allow for searching and quickly culling bad shots right in lightroom. 

Let me know if you want to collab on that.

**Comment by ajamess** [comment]
I had a few questions for you regarding model training

1) How did you train these models? On what datasets?
2) How hard would it be to train on a wider array of bird species? I notice that it will get "close" on speciation on the same bird, but often label the same bird as two different slightly similar looking species.
3) Could species detection be easily extended to a wider range of wildlife?
4) Could the quality model be genericized to detect focus quality on other subjects, such as human eyes?
5) If the quality model can be genericized, I believe this plugin could be useful for wedding photographers who want to cull out of focus / low quality images. Do you know of any good open eye / gaze detection models that could be added?

**Comment by SanjaySoniLV** [comment]
> Most of it is vibe-coded, so please don't skewer me on code quality lmao.

Oh don't worry. The GUI of project kestrel was mostly vibe coded too.

> How did you train these models? On what datasets?

Data for the species classifier was obtained using bing image search + public domain filtering. It was trained a long time ago, using an Azure Custom Vision pipeline that I no longer have access to. It's definitely not ideal. The training data is imperfectly labeled, and it used a different bird detection algorithm that returned bounding boxes instead of masks. However, it's the best I've got unless there are open-source versions of what Cornell Lab uses in Merlin.

The quality estimation model uses a dataset consisting of ~3000 photos from **my own photography work**. (I'm very proud of this 😄.) I took 3000 randomly selected images from my ~1.4TB life's work of photography, ran bird detection, then hand-classified them as good or bad. Then trained a very small convolutional neural network as a binary classifier to get the final quality score. Shockingly it works... that too quite well. I suspect there is a _lot_ that can be done to improve the quality classifier further (data augmentation strategies, etc.), but I'm training on my laptop CPU here LOL. 
 
> How hard would it be to train on a wider array of bird species? I notice that it will get "close" on speciation on the same bird, but often label the same bird as two different slightly similar looking species.

Unfortunately, I don't have access to good training hardware anymore. Species classification is notoriously difficult even for experienced birders for some species. I'm leaning towards a bucketing approach. Aka... if it guesses 3 different sparrows with low-confidence each, but the aggregate probability of all 'sparrow species' is high, just call the bird "Sparrow sp." 

I haven't had a chance to develop this bucketing approach too much, but it could potentially be done using the scientific hierarchical taxa classification that birds have.

Expanding to a wider array of bird species will be challenging. Probably best to have models trained for specific regions, like North America, South America, Europe, Asia, etc. I'm based in US, so I haven't explored any of these other regions at all. But on top of not having good training datasets, I don't have access to training hardware, so I'd be reliant on any open-source projects that are available for non-commercial use.

> Could species detection be easily extended to a wider range of wildlife?

This is also tough. Internally, Kestrel uses the Mask-RCNN model to identify which pixels correspond to a bird within the image. It just so happens that the Mask-RCNN project has a 'bird' category (the other categories feel quite random, like 'potted plant'). To expand to other species, we'd need a masking model that recognizes other wildlife. The low-hanging fruit here is to look at other categories of wildlife (like 'bear' or 'mammal') that are already trained into Mask-RCNN and implement those. Expanding beyond that will be tough.

> Could the quality model be genericized to detect focus quality on other subjects, such as human eyes?

This is certainly possible. However, there are quite a few existing 'image culling' projects out there that are focused on human subjects. I personally am a bird and wildlife photographer, and this project was born out of the lack of culling solutions for wildlife and birds.

> If the quality model can be genericized, I believe this plugin could be useful for wedding photographers who want to cull out of focus / low quality images. Do you know of any good open eye / gaze detection models that could be added?

I haven't looked into it too much. I know Mask-RCNN has some 'person' masks, and I'm sure there are pose estimators and eye detectors out there. The hardest part would be getting a lot of training data. Need thousands of high resolution (preferably RAW) files.

**Comment by ajamess** [comment]
Awesome response. Thank you!. I have about 700k images in my library, all types of RAWs back since 2006 or so. I probably have 50k wildlife shots in there. My first goal is to get the plugin functional, but maybe we can collab a bit once it get it to that point to see if my data set will be helpful. Your bucketing idea is a great one. It definitely sound complex and way above my expertise level! I am a lowly product guy by training.

I also have a Ryzen 7950x + 7900XTX 16GB + 192 GB desktop with 32TB of NVME storage and 100TB of attached storage at home, so I can reasonably train stuff, I just have never done it.

I am traveling for the next couple weeks, but when I get back, I will reach out :).

**Comment by SanjaySoniLV** [comment]
Sounds good! I'm sure theres a way to run ML training on AMD GPUs, but unfortunately I was unable to do it on my laptop. Probably easier with a dedicated card though.

**Comment by ajamess** [comment]
Doing a bit of research. 

I find the most painful step is actually focus detection, with the species detection being a nice bonus. It seems you are doing edge detection on the bird and using that for a quality rating. Do you think a dataset like this could be used for training a wider model to support focus detection on a more diverse set of creatures?

https://github.com/WildlifeDatasets/wildlife-datasets

**Comment by SanjaySoniLV** [comment]
Yeah, species detection is a nice bonus, and it doesn't really work _all_ that well right now anyway. 

The quality rating is not just edge detection. Edge detection is simply one of a few preprocessing steps that are run before feeding the resulting output into a machine learning model that scores the quality from 0 to 1. 

The machine learning model itself is quite small, so I wouldn't be surprised if it generalizes to other wildlife images without any modifications. In other words: I suspect that we wouldn't need to train a new model for different types of wildlife, because the idea of what's 'good quality' vs 'bad quality' doesn't really depend on what species of wildlife you're looking at.

However, a super crucial preprocessing step requires that we create an object mask. This is what that mask looks like:

<img width="995" height="346" alt="Image" src="https://github.com/user-attachments/assets/99a1a431-2cd4-411b-b949-7b8062638ad6" />

Kestrel uses Mask-RCNN (open source model that is easily instantiated with tensorflow) to create this bird mask. Mask-RCNN supports various categories right now, such as 'bird' and 'dog' and 'person' and so on - so there's low-hanging opportunity to just modify my code to integrate these already-supported categories without changing which machine learning models are used for the processing. I do plan to implement this in the future. 

The reason masking is ultra-critical is because it allows Kestrel to precisely differentiate subject from background when computing quality score. In previous versions of this project (see Project DoveEye if you're curious in my github, though it's horribly documented), I used an object detector which returned bounding boxes. The issue was, if the camera focused on a tree branch that was right in front of the bird, the algorithm would think, "Hey, look! something is nicely in focus! It must me good!" Switching to a masking approach + a machine learning approach made Kestrel MUCH better at capturing all the little subjective nuances we feel when we think something is 'high' or 'low' quality

Hope this made sense? 

Anyway... to your question: if we wanted to generalize quality detection to _all_ wildlife, we'd need two things:
1) a masking model that can detect which pixels belong to wildlife vs. background.
2) the quality scoring model to work well on non-bird wildlife.

For #1... the **easiest** option is to generalize to what Mask-RCNN already supports. The **best** option is to find a completely new, wildlife-focused masking model. 

For #2... my intuition is that it will work out of the box. If not, after implementing #1, I can look through some of my wildlife photos (or you could send some of yours) to re-train an algorithm with slightly more diverse data. But I highly doubt this is needed since the model is so small.

**Comment by SanjaySoniLV** [comment]
With the release of GPT-5, I've been messing around with things... LOL. So far I have planned a significant update to the visualizer that is way faster and allows the user to set their own star ratings, merge scenes together, and add scene names.

After I release that, I'm going to turn my attention towards generalizing the masking behavior to make full use of what Mask-RCNN supports. In theory it could help detect some more wildlife species which should make the tool a bit more useful - particularly for your WildlifeAI photoshop plugin.

And after that, I am going to see if I can integrate with EXIFTool or another solution to apply the ratings directly to the images themselves, either as a separate .xmp or within the image's own metadata. This is another piece of low-hanging fruit to integrate Kestrel into photoshop/darktable/etc. Since I don't have photoshop, perhaps you could help me test/develop this rating application feature?

**Comment by ajamess** [comment]
Thank you for the detailed response. I am back at home now after my vacation, so I am going to start iterating again on the Lightroom plugin now that I do not face the build issues I faced on my laptop.

Thank you for explaining the details behind your implementation, that is very helpful. Your ideas to expand into a more general tool are intriguing. I wonder if you would need more training data for the masking of different wildlife species? I wonder if various wildlife photogs would be willing to submit sample images for training a masking model, so long as the results retain their open source characteristics?

I am wondering, do you want to try to connect via email or some other form of collab tool? I would be interested in continuing this conversation in something a bit more fluid than github issue comments, if you are keen.

**Comment by SanjaySoniLV** [comment]
@NiyaNagi I've implemented XMP metadata writing. Can you try it out and see if it works? You should get the latest version from the compiled releases - the readme on this repo is very out of date

**Comment by SanjaySoniLV** [comment]
Closing this issue since XMP Metadata writing and lightroom integration seems to work well now!
